mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/generation/GeneratedCodePatterns.java

```diff
@@ -259,6 +292,113 @@ public static void setValue(
     }
   }
 
+  private static void generateParameterizedSet(
+      CodeBlock cqlName,
+      CodeBlock valueExtractor,
+      String targetName,
+      MethodSpec.Builder methodBuilder,
+      TypeName typeName,
+      BindableHandlingSharedCode enclosingClass,
+      CodeBlock nullSavingStrategy) {
+    generateSetWithNullSavingStrategy(
+        cqlName,
+        valueExtractor,
+        targetName,
+        methodBuilder,
+        nullSavingStrategy,
+        CodeBlock.of(
+            "$1L = $1L.set($2L, $3L, $4L)",
+            targetName,
+            cqlName,
+            valueExtractor,
+            enclosingClass.addGenericTypeConstant(typeName)));
+  }
+
+  private static void generateSetWithClass(
+      CodeBlock cqlName,
+      CodeBlock valueExtractor,
+      String targetName,
+      MethodSpec.Builder methodBuilder,
+      TypeName typeName,
+      CodeBlock nullSavingStrategy) {
+
+    generateSetWithNullSavingStrategy(
+        cqlName,
+        valueExtractor,
+        targetName,
+        methodBuilder,
+        nullSavingStrategy,
+        CodeBlock.of(
+            "$1L = $1L.set($2L, $3L, $4T.class)", targetName, cqlName, valueExtractor, typeName));
+  }
+
+  private static void generatePrimitiveSet(
+      CodeBlock cqlName,
+      CodeBlock valueExtractor,
+      String targetName,
+      MethodSpec.Builder methodBuilder,
+      String primitiveAccessor,
+      CodeBlock nullSavingStrategy) {
+
+    // todo how to handle primitive type getters? (int id == null is not possible)
+    //    generateSetWithNullSavingStrategy(cqlName, valueExtractor, targetName, methodBuilder,
+    // nullSavingStrategy,
+    //        CodeBlock.of(
+    //            "$1L = $1L.set$2L($3L, $4L)", targetName, primitiveAccessor, cqlName,
+    // valueExtractor));
+    methodBuilder.addStatement(
+        "$1L = $1L.set$2L($3L, $4L)",
+        targetName,
+        primitiveAccessor,
+        cqlName,
+        valueExtractor); // ignore nullSaving strategy for now
+  }
+
+  private static void generateSetWithNullSavingStrategy(
+      CodeBlock cqlName,
+      CodeBlock valueExtractor,
+      String targetName,
+      MethodSpec.Builder methodBuilder,
+      CodeBlock nullSavingStrategy,
+      CodeBlock nonNullStatement) {
+    if (nullSavingStrategy.equals(DO_NOT_SET_CODE_BLOCK)) {
+      methodBuilder.beginControlFlow("if($1L == null)", valueExtractor);
+      methodBuilder.addComment("do not set this field");
+      methodBuilder.endControlFlow();
+      methodBuilder.beginControlFlow("else");
```

Comment:
Another thought: if we leave values unset, this will raise an error:

```
java.lang.IllegalStateException: Unset value at index 1. If you want this value to be null, please set it to null explicitly.
```

Fortunately, we never send the query to the server, so I think that's how we want this to behave.

But I think we don't have any exposure to this in our testing because all mapper tests that use this option have a `@CassandraRequirement(min="3.4", description = "Creates a SASI index")` annotation.  We should probably reorg our tests so we can run as much of it on older versions as possible.  I think this would touch a lot of classes, so we should probably do this towards the end of development in this phase.

Reply:
Yes, I think that `java.lang.IllegalStateException: Unset value at index 1. If you want this value to be null, please set it to null explicitly.` is expected in this case.
I am +1 on extending our tests for Cassandra 2, 3 and 4. 
I think we should create a separate ticket to reorg our tests and cover 2 and 4 as well.
Do you want to create such a ticket or should I do it?

Reply:
I've created a JIRA ticket: https://datastax-oss.atlassian.net/browse/JAVA-2266